working/1426-extension-types/feature-specification.md

```diff
@@ -0,0 +1,948 @@
+qualified identifier denoting a generic type, but no actual type
+arguments*). In this case the omitted type arguments are determined by the
+corresponding superinterface of the on-type.
+
+*For example:*
+
+```dart
+extension type MyInt on int show num, isEven hide floor {
+  int get twice => 2 * this;
+}
+
+void main() {
+  MyInt m = 42;
+  m.twice; // OK, in the extension type.
+  m.isEven; // OK, a shown instance member.
+  m.ceil(); // OK, a shown instance member.
+  m.toString(); // OK, an `Object?` member.
+  m.floor(); // Error, hidden.
+}
+```
+
+
+### Implementing superinterfaces
+
+This section specifies the effect of having an `<interfaces>` part in an
+extension declaration.
+
+Let `E` be an extension declaration where `<interfaces>?` is of the form
+`implements T1, .. Tm`. We say that `E` has `T1` .. `Tm` as its direct
+superinterfaces.
+
+A compile-time error occurs if a direct superinterface does not denote a
+class, or if it denotes a class which cannot be a superinterface of a
+class.
+
+*For instance, `implements int` is an error.*
+
+If the `<interfaces>?` part of `E` is empty, it is treated as
+`implements Object`.
+
+For each member `m` named `n` in each direct superinterface of `E`, an
+error occurs unless `E` declares a member `m1` named `n` which is a correct
+override of `m`, or the show/hide part of `E` enables an instance member of
+the on-type which is a correct override of `m`.
+
+No subtype relationship exists between `E` and `T1, .. Tm`.
+
+*This means that when an extension type implements a set of interfaces, it
+is enforced that all the specified members are available, and that they
+have a signature which is compatible with the ones in `T1, .. Tm`, but
+there is no assignability from an expression of type `E` to a variable
+whose declared type is `Tj` for some `j` in 1..m. For that, it is necessary
+to use `box`, as described below.*
+
+
+### Gaining control over the instances: Protected extension types
+
+This section specifies the effect of including the keyword `protected` as
+the first token in an extension declaration.
+
+*The core idea is that no object can have a protected extension type as its
+static type unless it has been returned by an extension type constructor.
+This allows developers to gain control over which instances get to have
+that type.*
+
+Let `D` be an explicit extension type declaration named `E` which is
+prefixed by `protected`. We say that it is a _protected extension type
+declaration_, and that it introduces a _protected extension type_.
+
+Let `E` be the name of a protected extension type declaration.
+The type `E` (or
+<code>E<T<sub>1</sub>, .. T<sub>k</sub>></code>
+if `E` is generic) is a proper subtype of `Object?` and a proper supertype
+of `Never`.
+
+*In contrast to non-protected extension types, the extension type has no
+subtype relationship with its on-type.*
+
+It is a compile-time error if a protected extension type `E` is the
```

Comment:
Not sure about these.  I can see the argument, but since you can get around it in other ways, I'm not sure it's worth the restriction.

Reply:
I'm currently writing up a number of design variants for protected extension types, to highlight the trade-offs, and in particular to emphasize that some variants will impose serious restrictions on the coding style used with this kind of type (e.g., it could be an error to use it as a type argument).